[UnifiedPDF] Start handling text widget annotations.

sammygill opened the pull request:
#### e0038b5b7ace4173ca4547fbbf9f4e124dec0675
<pre>
[UnifiedPDF] Start handling text widget annotations.
<a href="https://bugs.webkit.org/show_bug.cgi?id=267967">https://bugs.webkit.org/show_bug.cgi?id=267967</a>
<a href="https://rdar.apple.com/problem/121474907">rdar://problem/121474907</a>

Reviewed by Simon Fraser.

To start handling text annotations, we start by handling mouse left
clicks. When doing this we check to see if there is an annotation under
the left click event and if is a text annotation widget. If this is the
case then we will set the active annotation to this one (making sure
to commit any other previously active annotation).

The PDFPluginAnnotation that is created will need the bounds for the
underlying annotation to properly style the HTML input element that is
used. This is a matter of getting the bounds from the annotation, which
is in page space, and then converting it into document space. Before
this patch the PDFPluginAnnotation was performing this conversion, but
I moved this responsibility to the PDFPlugins instead so that the
annotations can just do everything in document space.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::PDFPlugin::boundsForAnnotation const):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginAnnotation.mm:
(WebKit::PDFPluginAnnotation::updateGeometry):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDocumentLayout.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDocumentLayout.mm:
(WebKit::PDFDocumentLayout::indexForPage const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::UnifiedPDFPlugin):
(WebKit::documentSpaceToPageSpaceTransform):
(WebKit::UnifiedPDFPlugin::convertFromDocumentToPage const):
(WebKit::UnifiedPDFPlugin::convertFromPageToDocument const):
(WebKit::UnifiedPDFPlugin::handleMouseEvent):
(WebKit::UnifiedPDFPlugin::boundsForAnnotation const):
(WebKit::UnifiedPDFPlugin::setActiveAnnotation):

Canonical link: <a href="https://commits.webkit.org/273456@main">https://commits.webkit.org/273456@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b5132612777a78943a4f70ee01269d03686e9fe8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/35319 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/14250 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/37438 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/38062 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31860 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/16621 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/11299 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/30733 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/35866 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/12034 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/31464 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/10562 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/10604 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/39308 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/32096 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/31907 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/36600 "Found 1 new test failure: fast/repaint/animation-after-layer-scroll.html (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10757 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/8674 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34623 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/12517 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8109 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/11281 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/11574 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1099,9 +1143,23 @@
+    return pageSpaceBounds;
 }
 
+
```
Nit: We can remove this line?

sammygill replied:
Yes! Done!

aprotyas reviewed (commented):
Looking good to me on a cursory review. Will take a closer look later

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
```diff
@@ -296,6 +296,37 @@ class PDFPluginBase : public ThreadSafeRefCounted<PDFPluginBase>, public WebCore
+    "    left: 0; "
+    "    right: 0; "
+    "    bottom: 0; "
+    "    display: -webkit-box; "
```
At some point we should stop using deprecated flexbox.

aprotyas replied:
Is `display: flex;` an equivalent replacement?

sammygill replied:
I don't believe it is completely equivalent in all scenarios and there are some differences. (e.g. usage of `-webkit-line-clamp`). But perhaps for our purposes it could be!

sammygill committed:
[UnifiedPDF] Start handling text widget annotations.
https://bugs.webkit.org/show_bug.cgi?id=267967
rdar://problem/121474907

Reviewed by Simon Fraser.

To start handling text annotations, we start by handling mouse left
clicks. When doing this we check to see if there is an annotation under
the left click event and if is a text annotation widget. If this is the
case then we will set the active annotation to this one (making sure
to commit any other previously active annotation).

The PDFPluginAnnotation that is created will need the bounds for the
underlying annotation to properly style the HTML input element that is
used. This is a matter of getting the bounds from the annotation, which
is in page space, and then converting it into document space. Before
this patch the PDFPluginAnnotation was performing this conversion, but
I moved this responsibility to the PDFPlugins instead so that the
annotations can just do everything in document space.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::PDFPlugin::boundsForAnnotation const):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginAnnotation.mm:
(WebKit::PDFPluginAnnotation::updateGeometry):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDocumentLayout.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDocumentLayout.mm:
(WebKit::PDFDocumentLayout::indexForPage const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::UnifiedPDFPlugin):
(WebKit::documentSpaceToPageSpaceTransform):
(WebKit::UnifiedPDFPlugin::convertFromDocumentToPage const):
(WebKit::UnifiedPDFPlugin::convertFromPageToDocument const):
(WebKit::UnifiedPDFPlugin::handleMouseEvent):
(WebKit::UnifiedPDFPlugin::boundsForAnnotation const):
(WebKit::UnifiedPDFPlugin::setActiveAnnotation):

Canonical link: https://commits.webkit.org/273456@main
